scalableminds/webknossos
# Allow to interpolate segments for faster annotation

- implements annotating every 2nd slice so that the missing slices are interpolated
  - only works in XY (I think, this is also fine for now since this is the highest quality, anyway)
  - only when labeling (not erasing) <-- also intended
  - details:
    - a signed distance transform is computed for slice 1 and 3
    - the mean of these distances is computed and values below 0 are labeled then
- types the mjs / V2 / V3 module "properly" (ignore whitespace when reviewing)
- refactors the volume saga a bit (the helper module is new and only consists of moved code)
- misc (type) fixes
- Changes default of `dynamicSpaceDirection` property to false to avoid confusion.
- add `volumeInterpolationAllowed` to task settings which is disabled by default to avoid that tasks are accidentally done with interpolation

### URL of deployed dev instance (used for testing):
- https://interpolatesegments.webknossos.xyz

### Steps to test:
- create a new annotation
- switch to brush or trace tool
- enable the interpolation feature (see new button in toolbar)
- label every 2nd slice
- every (2nd+1) slice should be filled out automatically
- task handling
  - create a task type with volume enabled and volume interpolation allowed (default should be disallowed)
  - also enable the recommended settings and set volume interpolation to true (should be listed in the default recommended settings)
  - create a new task with the type from above
  - get task in dashboard
  - ensure that interpolation feature is enabled due to the recommendation and that it can be used
  - edit task type to allow interpolation
  - reload task annotation from above
  - ensure that interpolation feature cannot be used


### Todo:
- [x] make it work with other directions (i.e., decreasing z)
- Potentially as a follow-up
  - [x] multi-mag support
    - [X] make zoomed out interpolation work in 2-2-1
    - [x] make interpolation in 2-2-1 work when annotation is restricted to 2-2-1 (still buggy, could be related to https://github.com/scalableminds/webknossos/issues/6138)
    - [x] make interpolation work in 4-4-2 (i.e., z is not 1)
  - [x] interpolation depth > 2
    - [X] implement support
    - ~[ ] enable in UI (I think, we need to think about the details here; for example: what happens if I set the interpolation depth to 4 and then only go 3 slices forward? the interpolation will then look 4 slices back which is suboptimal for multiple reasons)~ I created https://github.com/scalableminds/webknossos/issues/6183 for this
  
### Issues:
- fixes #5434 

------
(Please delete unneeded items, merge only when none are left open)
